src/index.js

```diff
@@ -1,390 +1,586 @@
-  }
+  while (lifecycle.length > 0) lifecycle.pop()()
 
-  function removeElement(parent, element, node) {
-    function done() {
-      parent.removeChild(removeChildren(element, node))
-    }
+  return nextNode
+}
 
-    var cb = node.attributes && node.attributes.onremove
-    if (cb) {
-      cb(element, done)
-    } else {
-      done()
+export var h = function(name, props) {
+  var node
+  var rest = []
+  var children = []
+  var length = arguments.length
+
+  while (length-- > 2) rest.push(arguments[length])
+
+  if ((props = props == null ? {} : props).children != null) {
+    if (rest.length <= 0) {
+      rest.push(props.children)
     }
+    delete props.children
   }
 
-  function patch(parent, element, oldNode, node, isSvg) {
-    if (node === oldNode) {
-    } else if (oldNode == null || oldNode.nodeName !== node.nodeName) {
-      var newElement = createElement(node, isSvg)
-      parent.insertBefore(newElement, element)
-
-      if (oldNode != null) {
-        removeElement(parent, element, oldNode)
+  while (rest.length > 0) {
+    if (isArray((node = rest.pop()))) {
+      for (length = node.length; length-- > 0; ) {
+        rest.push(node[length])
       }
-
-      element = newElement
-    } else if (oldNode.nodeName == null) {
-      element.nodeValue = node
+    } else if (node === false || node === true || node == null) {
     } else {
-      updateElement(
-        element,
-        oldNode.attributes,
-        node.attributes,
-        (isSvg = isSvg || node.nodeName === "svg")
-      )
+      children.push(typeof node === "object" ? node : createTextVNode(node))
+    }
+  }
 
-      var oldKeyed = {}
-      var newKeyed = {}
-      var oldElements = []
-      var oldChildren = oldNode.children
-      var children = node.children
+  return typeof name === "function"
+    ? name(props, (props.children = children))
+    : createVNode(name, props, children, null, props.key, DEFAULT)
+}
 
-      for (var i = 0; i < oldChildren.length; i++) {
-        oldElements[i] = element.childNodes[i]
+var cancel = function(sub) {
+  sub.cancel()
+}
 
-        var oldKey = getKey(oldChildren[i])
-        if (oldKey != null) {
-          oldKeyed[oldKey] = [oldElements[i], oldChildren[i]]
-        }
+var update = function(sub, oldSub, dispatch) {
+  if (sub.effect === oldSub.effect) {
+    for (var k in merge(sub.props, oldSub.props)) {
+      if (sub.props[k] !== oldSub.props[k]) {
+        cancel(oldSub)
+        return start(sub, dispatch)
       }
+    }
+  }
+  return oldSub
+}
 
-      var i = 0
-      var k = 0
+var start = function(sub, dispatch) {
+  return merge(sub, {
+    cancel: sub.effect(sub.props, dispatch)
+  })
+}
 
-      while (k < children.length) {
-        var oldKey = getKey(oldChildren[i])
-        var newKey = getKey((children[k] = resolveNode(children[k])))
+var listen = function(sub, oldSub, dispatch) {
+  if (isArray(sub) || isArray(oldSub)) {
+    var out = []
+    var subs = isArray(sub) ? sub : [subs]
```

Comment:
Here you are doing duplicate isArray check for sub which is already done in if condition. So the tradeoff is having to variables to store result of isArray for sub and oldSub versus calling isArray two times. I will consider more variables as cheaper than more function call in runtime. My idea is something like this:
```js
const isArraySub = isArray(sub)
const isArrayOldSub = isArray(oldSub)
if (isArraySub || isArrayOldSub) {
  let out = []
  let subs = isArraySub ? sub : [subs];
  ...
}
```

Reply:
Well, that won't work with `let`, only with var. Var hoists the definition to the top of the scope, so when it's referenced at `[subs]` it returns `undefined`. But with `let` it will throw a ReferenceError, since it is self referencing and won't exist until the line has executed. I actually think this is a typo and you were intending to ensure that `sub` is an array. If that's the case, you can just do:
```
var subs = [].concat(sub)
```
Which will ensure that `subs` ends up an array and works if `sub` is already and array or just a value.

Reply:
@TheNando I have already made this suggestions one month ago on Slack 
see https://hyperapp.slack.com/archives/C5D86NQ3X/p1529772214000096?thread_ts=1529772214.000096&cid=C5D86NQ3X

If @jorgebucaran made the decision to do not use this implementation, he should have some good reasons.

Reply:
@TheNando I have already made this suggestions one month ago on Slack 
see https://hyperapp.slack.com/archives/C5D86NQ3X/p1529772214000096?thread_ts=1529772214.000096&cid=C5D86NQ3X

If @jorgebucaran made the decision to do not use this implementation, he should have some good reasons.

Reply:
Thanks, @Swizz. Basically `[].concat` is a lot slower.

Reply:
@jorgebucaran But `: [subs]` supposed to be `: [sub]`, right?

Reply:
@TheNando Yes, totally, that's a bug! Fixing it now :)